Sources/STARSDK/Cryptography/IdentityManager.swift

```diff
@@ -0,0 +1,163 @@
+//  Created by Loïc Gardiol on 30.03.20.
+
+import Foundation
+import CommonCrypto
+
+
+/// Responsible for generating the secret keys (SKs) and deriving the Ephemeral IDs (EphID)
+/// Currently not thread-safe.
+class IdentityManager {
+    
+    static let shared = IdentityManager()
+    
+    init() {}
+
+    /// Last nbStoredEpochs SKs, in chronological order. Last is current.
+    /// Generated on first access, then persisted securely and rotated.
+    /// Guarenteed to be non-empty
+    private var SKs: [SK] {
+        let persistenceKey = "SKs"
+        var sks: [SK] = {
+            if let persisted: [SK] = SecurePersistence.shared.object(key: persistenceKey) {
+                return persisted
+            }
+            #warning("TODO If Keychain access fails (device just rebooted for example) and we can't read the previously persisted keys, we are going to override them here.")
+            let initialSK = SK(epoch: Epoch.current, keyData: Utils.generateRandomKey(nbBits: GlobalParameters.SKNbBits))
+            return [initialSK]
+        }()
+        let currentEpoch = Epoch.current
+        if let last = sks.last, last.epoch != currentEpoch {
+            var sk = last
+            while sk.epoch < currentEpoch {
+                sk = sk.next
+                sks.append(sk)
+            }
+        }
+        if sks.count > GlobalParameters.nbStoredEpochs {
+            sks = Array(sks.dropFirst(sks.count - Int(GlobalParameters.nbStoredEpochs)))
+        }
+        SecurePersistence.shared.set(sks, key: persistenceKey)
+        return sks
+    }
+    
+    private var currentEphIDCache: EphID?
+    
+    var currentEphID: EphID {
+        let currentSK = self.SKs.last! // doc says so
+        if let cached = self.currentEphIDCache, cached.epoch == currentSK.epoch {
+            return cached
+        }
+        let broadcast = "broadcast".data(using: .utf8)!
+        let hmac = Utils.hmacSHA256(msg: broadcast, key: currentSK.keyData)
+        let truncated = hmac.dropLast(hmac.count-GlobalParameters.EphIDNbBytes).withUnsafeBytes { Data($0) }
+        let ephId = EphID(epoch: currentSK.epoch, data: truncated)
+        self.currentEphIDCache = ephId
+        return ephId
+    }
+}
+
+struct SK: Codable, CustomStringConvertible {
```

Comment:
can we rename SK to SecretKey for better code readability

Reply:
Sure. Use `SK` because this is the given name in the EPFL spec.